ambry-named-mysql/src/main/java/com/github/ambry/named/MySqlNamedBlobDb.java

```diff
@@ -928,42 +952,118 @@ private DeleteResult run_delete_v2(String accountName, String containerName, Str
     return new DeleteResult(blobVersions);
   }
 
-  private List<StaleNamedBlob> runPullStaleBlobs(final Connection connection) throws Exception {
-    String query = "";
-    try (PreparedStatement statement = connection.prepareStatement(GET_STALE_QUERY)) {
-      Calendar calendar = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
-      calendar.add(Calendar.DATE, -config.staleDataRetentionDays);
-      long previousTimeInMillis = calendar.getTimeInMillis();
-      statement.setLong(1, previousTimeInMillis * VERSION_BASE);
-      statement.setInt(2, config.queryStaleDataMaxResults);
-      query = statement.toString();
-      logger.debug("Pulling stale blobs from MySql. Query {}", query);
-      try (ResultSet resultSet = statement.executeQuery()) {
-        List<StaleNamedBlob> resultList = new ArrayList<>();
-        while (resultSet.next()) {
-          short accountId = resultSet.getShort(1);
-          short containerId = resultSet.getShort(2);
-          String blobName = resultSet.getString(3);
-          String blobId = Base64.encodeBase64URLSafeString(resultSet.getBytes(4));
-          long version = resultSet.getLong(5);
-          Timestamp currentTime = resultSet.getTimestamp(6);
-          StaleNamedBlob result = new StaleNamedBlob(accountId, containerId, blobName, blobId, version, currentTime);
-          resultList.add(result);
+  public static List<StaleNamedBlob> getStaleBlobs(List<StaleNamedBlob> blobList) {
+    List<StaleNamedBlob> staleBlobs = new ArrayList<>();
+    if (blobList.isEmpty()) {
+      return staleBlobs;
+    }
+    StaleNamedBlob keepBlob = blobList.get(0);
+    long cutoffTime = System.currentTimeMillis() - fiveDaysMillis;
+    Timestamp cutoffTimestamp = new Timestamp(cutoffTime);
+
+    for (int i = 1; i < blobList.size(); i++) {
+      StaleNamedBlob currentBlob = blobList.get(i);
+      if (currentBlob.getDeleteTs() != null) {
+        continue;
+        }
+
+      if (!keepBlob.getBlobName().equals(currentBlob.getBlobName())) {
+        keepBlob = currentBlob;
+        continue;
+      }
+
+      int keepBlobState = keepBlob.getBlobState();
+      int currentBlobState = currentBlob.getBlobState();
+      Timestamp keepBlobModifiedTS = keepBlob.getModifiedTS();
+
+      if (keepBlobState == 0 && currentBlobState == 1) {
+        if (keepBlobModifiedTS.before(cutoffTimestamp)) {
+          staleBlobs.add(keepBlob);
+          keepBlob = currentBlob;
+        }
+      } else if (keepBlobState == 1 && currentBlobState == 1) {
+        staleBlobs.add(currentBlob);
+      } else if (keepBlobState == 0 && currentBlobState == 0) {
+        if (keepBlobModifiedTS.after(cutoffTimestamp)) {
+          staleBlobs.add(currentBlob);
+        } else {
+          staleBlobs.add(keepBlob);
```

Comment:
let's not use else cause we might add new stats in the future. it's better we can specify the exact condition.